### PR TITLE
TorchTrainerを実装

### PR DESCRIPTION
**Pull request by Geson-anko**
# Pull Request

## 内容

TorchTrainerを実装しました。`optimizers`や`lr_schedulers`に関してはサブクラスに隠蔽しておく必要性が低いと感じたため、private 属性 `_` prefixを除外しました。

#149 のやつは私が書き直したのでPRを送り直しました。

## 他のコード・機能への影響

- [x] なし
- [ ] あり

## Submit前の確認項目

- [x] タイトルは一目でわかるようにし、説明文は PR を簡潔に説明するようにしましたか?
- [x] あなたの PR が、異なる変更を束ねたものではなく、ひとつのことだけを行うものであることを確認しましたか?
- [x] この PR で導入されたすべての変更点を記述、リストアップしましたか?
- [x] PR を`make run`コマンドでローカルでテストしましたか?

### CODE REVIEW

File: src/pamiq_core/torch/trainer.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        """
+        super().__init__(
+            training_condition_data_user,
+            min_buffer_size,
+            min_new_data_count,
+        )
+
+        # Containers for optimizer and scheduler instances
+        self.optimizers: OptimizersDict = {}
+        self.lr_schedulers: LRSchedulersDict = {}
```
Comment (zassou65535) [inline src/pamiq_core/torch/trainer.py:66]: `self.schedulers`としてしまった方が`self.optimizers`と文字数が揃うからより見通しをよくできる気がする
```suggestion
        self.schedulers: LRSchedulersDict = {}
```

File: src/pamiq_core/torch/trainer.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+        # Load optimizer states from disk
+        for optimizer_path in path.glob("*.optim.pt"):
+            name = optimizer_path.name.replace(".optim.pt", "")
+            self.optimizer_states[name] = torch.load(optimizer_path)  # pyright: ignore[reportUnknownMemberType]
+
+        # Load scheduler states from disk
+        for scheduler_path in path.glob("*.lrsch.pt"):
+            name = scheduler_path.name.replace(".lrsch.pt", "")
+            self.lr_scheduler_states[name] = torch.load(scheduler_path)  # pyright: ignore[reportUnknownMemberType]
```
Comment (zassou65535) [inline src/pamiq_core/torch/trainer.py:224]: `load_states`って名前なのに実行するとモデルには実体にパラメーターが読み込まれて、最適化関数とスケジューラーはstatesに読み込まれるだけで実体には読み込まれないの直感的じゃない気がする

File: src/pamiq_core/torch/trainer.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        """
+        super().save_state(path)
+
+        # Save optimizer states to disk
+        for name, optimizer_state in self.optimizer_states.items():
+            torch.save(optimizer_state, path / f"{name}.optim.pt")  # pyright: ignore[reportUnknownMemberType]
+
+        # Save scheduler states to disk
+        for name, scheduler_state in self.lr_scheduler_states.items():
+            torch.save(scheduler_state, path / f"{name}.lrsch.pt")  # pyright: ignore[reportUnknownMemberType]
```
Comment (zassou65535) [inline src/pamiq_core/torch/trainer.py:196]: 一度`teardown`を実行しないと`save_state`で正しく保存できないのなんか気になる　`teardown`ってモデル全部の情報を消去するようなニュアンスがあるけど、この実装だとその後に`save_state`ができてしまうし、なんならそうしなければならないの非自明だと思った
関数の頭で`self._keep_optimizer_and_scheduler_states()`を実行しておくといいかも？
```suggestion
    def save_state(self, path: Path) -> None:
        """Save trainer state to disk.

        Persists the states of all optimizers and schedulers to the specified
        directory path.

        Args:
            path: Directory path where state should be saved
        """
        super().save_state(path)
        
        # Prepare parameters to save
        self._keep_optimizer_and_scheduler_states()

        # Save optimizer states to disk
        for name, optimizer_state in self.optimizer_states.items():
            torch.save(optimizer_state, path / f"{name}.optim.pt")  # pyright: ignore[reportUnknownMemberType]

        # Save scheduler states to disk
        for name, scheduler_state in self.lr_scheduler_states.items():
            torch.save(scheduler_state, path / f"{name}.lrsch.pt")  # pyright: ignore[reportUnknownMemberType]
```

File: src/pamiq_core/torch/trainer.py (hunk 1 above)
Comment (Geson-anko) [reply src/pamiq_core/torch/trainer.py:66]: `pamiq-core`ではシステムのイベント処理をスケジューリングするためにも`Scheduler`という言葉を用いているので、その重複を避けるために`lr_`をつけている

File: src/pamiq_core/torch/trainer.py (hunk 2 above)
Comment (Geson-anko) [reply src/pamiq_core/torch/trainer.py:224]: @zassou65535 `setup`処理で読み込まれるので良いのです。

File: src/pamiq_core/torch/trainer.py (hunk 3 above)
Comment (Geson-anko) [reply src/pamiq_core/torch/trainer.py:196]: 学習後の状態が存在しない場合は不要な保存処理になるので無駄。最小のコードで必要な機能を実現している。

File: src/pamiq_core/torch/trainer.py (hunk 1 above)
Comment (zassou65535) [reply src/pamiq_core/torch/trainer.py:66]: 名前の重複があるなら確かに話は別になるかも　`lr_`あった方がよさそう

File: src/pamiq_core/torch/trainer.py (hunk 2 above)
Comment (zassou65535) [reply src/pamiq_core/torch/trainer.py:224]: ということは`load_states`の後に`setup`が呼び出される想定？じゃあ処理の上では問題なさそう
(とはいえpytorchではモデルにしろ最適化関数にしろ`load_states`を読んだらほぼ絶対実体にパラメーターが読み込まれるはずだし、pamiq-coreでもここ以外では実体に読み込まれてるのにここだけそうじゃないの、初見の人ここ絶対ハマりポイントになるからもしそうするならコメントが欲しい)

File: src/pamiq_core/torch/trainer.py (hunk 3 above)
Comment (zassou65535) [reply src/pamiq_core/torch/trainer.py:196]: これも上に同じく、もしこれで行くならコメントが欲しい

Comment (Geson-anko) [reply src/pamiq_core/torch/trainer.py]: コメント追加 https://github.com/MLShukai/pamiq-core/pull/177/commits/8eabd140a69b1fc3b54af9b1087e7c7954d23b95

File: src/pamiq_core/torch/trainer.py (hunk 2 above)
Comment (Geson-anko) [reply src/pamiq_core/torch/trainer.py:226]: コメント追加: https://github.com/MLShukai/pamiq-core/pull/177/commits/8eabd140a69b1fc3b54af9b1087e7c7954d23b95

File: src/pamiq_core/torch/trainer.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        for optimizer_path in path.glob("*.optim.pt"):
+            name = optimizer_path.name.replace(".optim.pt", "")
+            self.optimizer_states[name] = torch.load(optimizer_path)  # pyright: ignore[reportUnknownMemberType]
+
+        # Load scheduler states from disk
+        for scheduler_path in path.glob("*.lrsch.pt"):
+            name = scheduler_path.name.replace(".lrsch.pt", "")
+            self.lr_scheduler_states[name] = torch.load(scheduler_path)  # pyright: ignore[reportUnknownMemberType]
+
+        # Load the state into the optimizer and LR scheduler within the `setup` method.
```
Comment (zassou65535) [inline src/pamiq_core/torch/trainer.py:228]:
```suggestion
        # After This method performed, load the state into the optimizer and LR scheduler by the `setup` method.
```

Comment (zassou65535) [reply src/pamiq_core/torch/trainer.py:228]: ちょっとだけ加筆

File: src/pamiq_core/torch/trainer.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+        Persists the states of all optimizers and schedulers to the specified
+        directory path.
+
+        Args:
+            path: Directory path where state should be saved
+        """
+        super().save_state(path)
+
+        # Keep the state of the optimizer or LR scheduler within the `teardown` method.
```
Comment (zassou65535) [inline src/pamiq_core/torch/trainer.py:190]: ちょっとだけ加筆したい
```suggestion
         # Before this method performed, the state of the optimizer or LR scheduler have already been kept by the `teardown` method.
```

**Review by zassou65535 — APPROVED** [review]
